File: Universal.Xaml.Advertising/AdPlayerFactoryPlugin.cs
using Microsoft.Media.Advertising;
using System;
using System.Linq;
using System.Collections.Generic;
#if SILVERLIGHT
using System.Windows;
#else
using Windows.UI.Xaml.Media;
using Windows.Graphics.Imaging;
using Windows.UI.Xaml;
#endif

namespace Microsoft.PlayerFramework.Advertising
{
    /// <summary>
    /// Provides a default implementation of IAdPlayerFactoryPlugin that is capable of serving linear video ads or nonlinear image ads.
    /// </summary>
    public class AdPlayerFactoryPlugin : IAdPlayerFactoryPlugin
    {
        /// <summary>
        /// Gets the collection MIME types supported for video ads.
        /// </summary>
        public IList<string> SupportedVideoMimeTypes { get; private set; }

        /// <summary>
        /// Gets or sets at which the point at which ads can be skipped.
        /// Note: This is used only if VAST 3.0 does not specify this already.
        /// Can be in TimeSpan format or percentage.
        /// Default = null
        /// </summary>
        public FlexibleOffset SkippableOffset { get; set; }

        /// <summary>
        /// Gets or sets if the plugin is enabled.
        /// </summary>
        public bool IsEnabled { get; set; }

        /// <summary>
        /// Gets or sets a custom style to be used for the VpaidVideoAdPlayer
        /// </summary>
        public Style VpaidVideoAdPlayerStyle { get; set; }

        /// <summary>
        /// Gets or sets a custom style to be used for the VpaidImageAdPlayer
        /// </summary>
        public Style VpaidImageAdPlayerStyle { get; set; }

        /// <summary>
        /// Gets or sets a custom style to be used for the VpaidWebAdPlayer (For IFrame and HTML nonlinear resources)
        /// </summary>
        public Style VpaidWebAdPlayerStyle { get; set; }

        /// <summary>
        /// Creates a new instance of AdPlayerFactoryPlugin
        /// </summary>
        public AdPlayerFactoryPlugin()
        {
            SupportedVideoMimeTypes = new List<string>(new[] { "video/x-ms-wmv", "video/mp4" });
            IsEnabled = true;
        }

        /// <inheritdoc /> 
        public IVpaid GetPlayer(ICreativeSource creativeSource)
        {
            if (!IsEnabled || creativeSource.MimeType == null) return null;
            var skippableOffset = creativeSource.SkippableOffset ?? SkippableOffset;
            if (creativeSource.Type == CreativeSourceType.Linear)
            {
                if (creativeSource.ApiFramework == "VPAID")
                {
                    //switch (creativeSource.MimeType)
                    //{
                    //    case "application/x-javascript":
                    //        return new VpaidJavaScriptAdPlayer(skippableOffset, creativeSource.Duration, creativeSource.ClickUrl) { Style = VpaidWebAdPlayerStyle };
                    //}
                }
                else
                {
#if SILVERLIGHT
                    if (SupportedVideoMimeTypes.Contains(creativeSource.MimeType.ToLowerInvariant()))
#else
                    if (SupportedVideoMimeTypes.Contains(creativeSource.MimeType.ToLowerInvariant()) || CanPlayCodec(creativeSource.Codec))
#endif
                    {
                        return new VpaidVideoAdPlayer(skippableOffset, creativeSource.Duration, creativeSource.ClickUrl) 
                        { 
                            Style = VpaidVideoAdPlayerStyle,
#if NETFX_CORE
                            AudioCategory = MediaPlayer.AudioCategory
#endif
                        };
                    }
                }
            }
            else if (creativeSource.Type == CreativeSourceType.NonLinear)
            {
                switch (creativeSource.MediaSourceType)
                {
                    case MediaSourceEnum.Static:
#if SILVERLIGHT
                        if (creativeSource.MimeType.ToLowerInvariant().StartsWith("image/"))
#else
                        if (BitmapDecoder.GetDecoderInformationEnumerator().SelectMany(d => d.MimeTypes).Select(m => m.ToLowerInvariant()).Contains(creativeSource.MimeType.ToLowerInvariant()))
#endif
                        {
                            return new VpaidImageAdPlayer(skippableOffset, creativeSource.Duration, creativeSource.ClickUrl) { Style = VpaidImageAdPlayerStyle };
                        }
                        else return null;
                    case MediaSourceEnum.IFrame:
                        return new VpaidIFrameAdPlayer(skippableOffset, creativeSource.Duration, creativeSource.ClickUrl, creativeSource.Dimensions) { Style = VpaidWebAdPlayerStyle };
                    case MediaSourceEnum.HTML:
                        return new VpaidHtmlAdPlayer(skippableOffset, creativeSource.Duration, creativeSource.ClickUrl, creativeSource.Dimensions) { Style = VpaidWebAdPlayerStyle };
                }
            }
            return null;
        }
#if !SILVERLIGHT
        private bool CanPlayCodec(string codec)
        {
            if (!string.IsNullOrEmpty(codec))
            {
                var canPlayCodecResponse = MediaPlayer.CanPlayType(codec);
                return canPlayCodecResponse == MediaCanPlayResponse.Maybe || canPlayCodecResponse == MediaCanPlayResponse.Probably;
            }
            return false;
        }
#endif

        void IPlugin.Load() { }

        void IPlugin.Update(IMediaSource mediaSource) { }

        void IPlugin.Unload() { }

        /// <inheritdoc /> 
        public MediaPlayer MediaPlayer { get; set; }
    }
}

File: Universal.Js.Advertising/js/AdPlayerFactoryPlugin.js
(function (PlayerFramework, undefined) {
    "use strict";

    // AdPlayerFactoryPlugin Errors
    var invalidConstruction = "Invalid construction: AdPlayerFactoryPlugin constructor must be called using the \"new\" operator.";

    // AdPlayerFactoryPlugin Class
    var AdPlayerFactoryPlugin = WinJS.Class.derive(PlayerFramework.Advertising.AdPlayerFactoryPluginBase, function (options) {
        if (!(this instanceof PlayerFramework.Plugins.AdPlayerFactoryPlugin)) {
            throw invalidConstruction;
        }

        PlayerFramework.PluginBase.call(this, options);
    }, {
        // Public Properties
        supportedVideoMimeTypes: {
            get: function () {
                return this._supportedVideoMimeTypes;
            },
            set: function (value) {
                this._supportedVideoMimeTypes = value;
            }
        },

        skippableOffset: {
            get: function () {
                return this._skippableOffset;
            },
            set: function (value) {
                this._skippableOffset = value;
            }
        },

        clickThruLinearText: {
            get: function () {
                return this._clickThruLinearText;
            },
            set: function (value) {
                this._clickThruLinearText = value;
            }
        },

        clickThruNonLinearText: {
            get: function () {
                return this._clickThruNonLinearText;
            },
            set: function (value) {
                this._clickThruNonLinearText = value;
            }
        },

        // Public Methods
        getPlayer: function (creativeSource) {
            if (creativeSource.type === Microsoft.Media.Advertising.CreativeSourceType.linear){
                if (creativeSource.apiFramework == "VPAID") {
                    // TODO: support application/x-javascript
                }
                else{
                    if (this._supportsMimeType(creativeSource.mimeType, this.supportedVideoMimeTypes) || this._canPlayType(creativeSource.codec)) {
                        var result = new PlayerFramework.Advertising.VpaidVideoAdPlayer(creativeSource.skippableOffset || Microsoft.Media.Advertising.FlexibleOffset.parse(this.skippableOffset), creativeSource.duration, creativeSource.clickUrl, this.clickThruLinearText);
                        result.msAudioCategory = this.mediaPlayer.msAudioCategory;
                        return result;
                    }
                }
            }
            else if (creativeSource.type === Microsoft.Media.Advertising.CreativeSourceType.nonLinear) {
                switch (creativeSource.mediaSourceType) {
                    case Microsoft.Media.Advertising.MediaSourceEnum.static:
                        if (this._supportsMimeType(creativeSource.mimeType, PlayerFramework.Utilities.getImageMimeTypes())) {
                            return new PlayerFramework.Advertising.VpaidImageAdPlayer(creativeSource.skippableOffset || Microsoft.Media.Advertising.FlexibleOffset.parse(this.skippableOffset), creativeSource.duration, creativeSource.clickUrl, this.clickThruNonLinearText, creativeSource.dimensions);
                        }
                    case Microsoft.Media.Advertising.MediaSourceEnum.iframe:
                        return new PlayerFramework.Advertising.VpaidIFrameAdPlayer(creativeSource.skippableOffset || Microsoft.Media.Advertising.FlexibleOffset.parse(this.skippableOffset), creativeSource.duration, creativeSource.clickUrl, this.clickThruNonLinearText, creativeSource.dimensions);
                    case Microsoft.Media.Advertising.MediaSourceEnum.html:
                        return new PlayerFramework.Advertising.VpaidHtmlAdPlayer(creativeSource.skippableOffset || Microsoft.Media.Advertising.FlexibleOffset.parse(this.skippableOffset), creativeSource.duration, creativeSource.clickUrl, this.clickThruNonLinearText, creativeSource.dimensions);
                }

            }
            return null;
        },

        // Private Methods
        _canPlayType: function(codec) {
            try {
                return this.mediaPlayer.canPlayType(codec);
            } catch (e) {
                // HACK: this will crash on some types due to bug in IE
                return false;
            }
        },

        _setOptions: function (options) {
            PlayerFramework.Utilities.setOptions(this, options, {
                isEnabled: true,
                skippableOffset: null,
                clickThruLinearText: PlayerFramework.Utilities.getResourceString("AdLinkLabel"),
                clickThruNonLinearText: "",
                supportedVideoMimeTypes: [
                    "video/mp4",
                    "video/x-ms-wmv"
                ]
            });
        },

        _supportsMimeType: function (mimeType, supportedMimeTypes) {
            return !!PlayerFramework.Utilities.first(supportedMimeTypes, function (supportedMimeType) {
                return supportedMimeType.toLowerCase() === mimeType.toLowerCase();
            });
        }
    });

    // AdPlayerFactoryPlugin Mixins
    WinJS.Class.mix(PlayerFramework.MediaPlayer, {
        adPlayerFactoryPlugin: {
            value: null,
            writable: true,
            configurable: true
        }
    });

    // AdPlayerFactoryPlugin Exports
    WinJS.Namespace.define("PlayerFramework.Plugins", {
        AdPlayerFactoryPlugin: AdPlayerFactoryPlugin
    });

})(PlayerFramework);

